src/experimental/tools/download2go/index.ts

```diff
@@ -0,0 +1,415 @@
+/**
+ * Copyright 2019 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IDBPDatabase } from "idb";
+import { AsyncSubject } from "rxjs";
+
+import EventEmitter from "../../../utils/event_emitter";
+import PPromise from "../../../utils/promise";
+
+import { IActiveDownload, IActivePauses } from "./api/context/types";
+import { setUpDb } from "./api/db/dbSetUp";
+import DownloadManager from "./api/downloader/downloadManager";
+import {
+  getBuilderFormattedForAdaptations,
+  getBuilderFormattedForSegments,
+  getKeySystemsSessionsOffline,
+  offlineManifestLoader,
+} from "./api/downloader/manifest";
+import { IContentProtection } from "./api/drm/types";
+import {
+  IContentLoader,
+  IDownload2GoEvents,
+  IEmitterTrigger,
+  IGlobalSettings,
+  IInitSettings,
+  IStoredManifest,
+} from "./types";
+import {
+  checkForPauseAMovie,
+  checkForResumeAPausedMovie,
+  checkInitDownloaderOptions,
+  IndexDBError,
+  isSupported,
+  SegmentConstuctionError,
+  ValidationArgsError,
+} from "./utils";
+
+/**
+ * Instanciate a D2G downloader.
+ * @param {Object<{nameDB, storeManifestEvery}>} IOptionsStarter
+ * @return {IPublicAPI} IPublicAPI
+ */
+class D2G extends EventEmitter<IDownload2GoEvents> {
+  static isPersistentLicenseSupported(): Promise<boolean> {
+    return isSupported();
+  }
+
+  public readonly nameDB: string;
+  private db: IDBPDatabase | null;
+  private emitter: IEmitterTrigger<IDownload2GoEvents>;
+  private activeDownloads: IActiveDownload;
+  private activePauses: IActivePauses;
+
+  constructor(options: IGlobalSettings = {}) {
+    super();
+    this.nameDB =
+      options.nameDB == null || options.nameDB === ""
+        ? "d2g-canalplus"
+        : options.nameDB;
+    this.activeDownloads = {};
+    this.activePauses = {};
+    this.db = null;
+    this.emitter = {
+      trigger: (eventName, payload) => this.trigger(eventName, payload),
+    };
+  }
+
+  /**
+   * Initialize an IndexDB instance.
+   * @returns {Promise<void>}
+   */
+  initDB(): Promise<IDBPDatabase> {
```

Comment:
Also, the user does not care that we rely on a sort of "database", no? We may even rely on more things in the future. Why not just `initialize`?

Reply:
Ha, It's will be tough to decide.

I like the idea of lazily load the database because we can get rid of the `initDB()` which deletes a step for the user.

However, if we lazy load the DB we will still have the DB in our API in private right?
Because, if the user launches a download thanks to the `download` method and then pause and resume again, we will not initialize again the DB since we already called the download method, so we already initialized the DB.

Concerning, a potential `initialize` function instead of `initDB`, I also agree, that could be useful in the future, do you have any ideas for what we can use `initialize` other than DB initializing?

Reply:
I might go for `initialize` logic since it seems between both solutions? what do you think?